crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -156,14 +156,17 @@ impl Fulfillment {
             Side::Sell => executed,
         };
         // Sell slightly more `sell_token` to capture the `surplus_fee`
-        let executed_sell_amount_with_fee = executed_sell_amount
-            .checked_add(
-                // surplus_fee is always expressed in sell token
-                self.surplus_fee()
-                    .map(|fee| fee.0)
-                    .ok_or(trade::Error::ProtocolFeeOnStaticOrder)?,
-            )
-            .ok_or(trade::Error::Overflow)?;
+        let executed_sell_amount_with_fee = match self.order().side {
+            Side::Buy => executed_sell_amount
+                .checked_add(
+                    // surplus_fee is always expressed in sell token
+                    self.surplus_fee()
+                        .map(|fee| fee.0)
+                        .ok_or(trade::Error::ProtocolFeeOnStaticOrder)?,
+                )
+                .ok_or(trade::Error::Overflow)?,
+            Side::Sell => executed_sell_amount,
```

Comment:
The reason is that the autopilot (or solver reward script to be more precise) is currently calculating volume based fees in surplus token, so that it can be compared to fee from surplus (which is always) in surplus token. I can try tomorrow to refactor this part of the code to use surplus token, I believe it will make more sense then.

Reply:
Thanks for raising this. I refactored the protocol fee to be calculated in surplus token (which is natural and already discussed with solver team internally as something we want to refactor).

Now, when applying protocol fee to `Fulfillment`, the conversion of the fee from surplus token to sell token is done at the very end and only because our `Fulfillment` object requires it. 
Also, I think now it's much easier to understand why we add surplus fee to BUY orders only, and it's because we want to consider the full amounts (from user point of view) when calculating volume based fee.